### PR TITLE
Feature/invoice location confirmation

### PR DESCRIPTION
**Pull request by hugolopespinto**
When the user rent an article, he will receive an email saying that the rent is confirmed, an invoice will also be saved in the database. The user can receive the invoice clicking in the button  dedicated to it, the button is in the confirm rent page and in the details of the location (accessible from the locations history).

coverage 68.9%
image

image
image
[facture_1.webm](https://github.com/H-chauvet/RISU/assets/72020673/26f63bc8-2514-4600-86c8-79844eb11af2)
[facture_2.webm](https://github.com/H-chauvet/RISU/assets/72020673/b3958e99-8484-42d4-bcbe-5c7842b72fd4)
image

### CODE REVIEW

**Review by invpuppet — CHANGES_REQUESTED** [review]

**Review by invpuppet — APPROVED** [review]

File: Application/mobile/app_code/lib/pages/rent/confirm/confirm_rent_state.dart [hunk 1]
```diff
@@ -8,10 +8,62 @@ import 'package:risu/pages/article/article_list_data.dart';
...
   late int hours;
   late ArticleData data;
+  final LoaderManager _loaderManager = LoaderManager();
+
+  void sendInvoice() async {
+    try {
+      setState(() {
+        _loaderManager.setIsLoading(true);
+      });
+      final response = await http.get(
```
Comment (31Nathan) [inline Application/mobile/app_code/lib/pages/rent/confirm/confirm_rent_state.dart:27]: get request for not getting anything ? Change the request to be coherent

File: Application/mobile/app_code/lib/pages/rent/return_state.dart [hunk 2]
```diff
@@ -33,6 +33,52 @@ class ReturnArticleState extends State<ReturnArticlePage> {
     },
   };
 
+  void sendInvoice() async {
+    try {
+      setState(() {
+        _loaderManager.setIsLoading(true);
+      });
+      final response = await http.get(
```
Comment (31Nathan) [inline Application/mobile/app_code/lib/pages/rent/return_state.dart:41]: Same here

File: Application/mobile/app_code/test/confirm_rent_test.dart [hunk 3]
```diff
@@ -18,13 +18,34 @@ void main() {
...
+      name: 'Ballon',
+      available: true,
+      price: 8,
+    );
     await tester.pumpWidget(initPage(ConfirmRentPage(hours: 5, data: data)));
     Finder homeButton = find.byKey(const Key('confirm_rent-button-back_home'));
     expect(homeButton, findsOneWidget);
 
-    await tester.tap(homeButton);
+    await tester.tap(homeButton, warnIfMissed: false);
```
Comment (31Nathan) [inline Application/mobile/app_code/test/confirm_rent_test.dart:31]: Not a good thing the "warnIfMissed"

File: Application/mobile/app_code/test/confirm_rent_test.dart [hunk 4]
```diff
@@ -18,13 +18,34 @@ void main() {
...
+      name: 'Ballon',
+      available: true,
+      price: 8,
+    );
+    await tester.pumpWidget(initPage(ConfirmRentPage(hours: 5, data: data)));
+    Finder invoiceButton =
+        find.byKey(const Key('rent_return-button-send_invoice'));
+    expect(invoiceButton, findsOneWidget);
+
+    await tester.tap(invoiceButton, warnIfMissed: false);
```
Comment (31Nathan) [inline Application/mobile/app_code/test/confirm_rent_test.dart:48]: Same

File: Application/mobile/app_code/test/rent_return_test.dart [hunk 5]
```diff
@@ -25,4 +25,17 @@ void main() {
...
   );
+  testWidgets(
+    'Click on the button to receive invoice',
+    (WidgetTester tester) async {
+      await tester.pumpWidget(initPage(const ReturnArticlePage(rentId: -1)));
+
+      Finder sendInvoiceButton =
+          find.byKey(const Key('rent_return-button-return_article'));
+      expect(sendInvoiceButton, findsNothing);
+      /*await tester.tap(sendInvoiceButton);
```
Comment (31Nathan) [inline Application/mobile/app_code/test/rent_return_test.dart:36]: Why push commented lines ?

File: Server/routes/Mobile/rent.js [hunk 6]
```diff
@@ -1,10 +1,140 @@
...
 
 const router = express.Router()
 const passport = require('passport')
 const rentCtrl = require("../../controllers/Mobile/rent")
 const userCtrl = require("../../controllers/Mobile/user")
 const itemCtrl = require("../../controllers/Common/items")
+const transporter = require('../../middleware/transporter')
+const containerCtrl = require('../../controllers/Common/container')
+
+function formatDate(date) {
```
Comment (31Nathan) [inline Server/routes/Mobile/rent.js:12]: Can be used in other file, create a specific file for it that can be imported from anywhere

File: Server/routes/Mobile/rent.js [hunk 7]
```diff
@@ -1,10 +1,140 @@
...
+  const year = date.getFullYear();
+  const hours = date.getHours();
+  const minutes = date.getMinutes();
+
+  const formattedDate = `${day}/${month}/${year} ${hours}:${minutes}`;
+
+  return formattedDate;
+}
+
+function drawTable(doc, tableData) {
```
Comment (31Nathan) [inline Server/routes/Mobile/rent.js:24]: Same here

File: Server/routes/Mobile/rent.js [hunk 8]
```diff
@@ -45,6 +217,39 @@ router.post('/article',
...
+
+      if (!location) {
+        return res.status(404).send('Location not found');
+      }
+
+      if (!location.invoice) {
+        return res.status(404).send('Invoice not found');
+      }
+
+      sendInvoice(location.invoice, user.email);
```
Comment (31Nathan) [inline Server/routes/Mobile/rent.js:243]: need to await

**Review by 31Nathan — CHANGES_REQUESTED** [review]
Missing tests for the back end part

File: Server/routes/Mobile/utils.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+  const tableHeight = tableData.length * (fontSize + cellPadding);
+
+  doc.rect(startX, startY, tableWidth, tableHeight).stroke();
+}
+
+module.exports = {
+    formatDate,
+    drawTable
+};
```
Comment (tbellicha) [inline Server/routes/Mobile/utils.js:36]: New line missing

**Review by tbellicha — CHANGES_REQUESTED** [review]

File: Server/routes/Mobile/utils.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
+function formatDate(date) {
```
Comment (invpuppet) [inline Server/routes/Mobile/utils.js:1]: no coverage directly on this file (the one on the branch come from the test on the invoice route)

File: Server/routes/Mobile/rent.js [hunk 11]
```diff
@@ -45,6 +185,39 @@ router.post('/article',
   }
 )
 
+router.put('/:locationId/invoice',
```
Comment (invpuppet) [inline Server/routes/Mobile/rent.js:188]: you only test the case where there is no error

**Review by invpuppet — CHANGES_REQUESTED** [review]

File: Application/mobile/app_code/lib/pages/rent/confirm/confirm_rent_state.dart [hunk 12]
```diff
@@ -8,10 +8,60 @@ import 'package:risu/pages/article/article_list_data.dart';
...
   late int hours;
   late ArticleData data;
+  final LoaderManager _loaderManager = LoaderManager();
+
+  void sendInvoice() async {
+    try {
+      setState(() {
+        _loaderManager.setIsLoading(true);
+      });
+      final response = await http.put(
```
Comment (tbellicha) [inline Application/mobile/app_code/lib/pages/rent/confirm/confirm_rent_state.dart:27]: Put is to update, not to create

File: Application/mobile/app_code/lib/pages/rent/return_state.dart [hunk 13]
```diff
@@ -33,6 +33,50 @@ class ReturnArticleState extends State<ReturnArticlePage> {
     },
   };
 
+  void sendInvoice() async {
+    try {
+      setState(() {
+        _loaderManager.setIsLoading(true);
+      });
+      final response = await http.put(
```
Comment (tbellicha) [inline Application/mobile/app_code/lib/pages/rent/return_state.dart:41]: Put is to update, not to create

File: Application/mobile/app_code/lib/pages/rent/confirm/confirm_rent_state.dart [hunk 14]
```diff
@@ -120,6 +170,17 @@ class ConfirmRentState extends State<ConfirmRentPage> {
                       ),
                     ),
                     const SizedBox(height: 32),
+                    SizedBox(
+                      width: double.infinity,
+                      child: MyOutlinedButton(
+                        text: AppLocalizations.of(context)!.receiveInvoice,
+                        key: const Key('return-rent-button-receive_invoice'),
```
Comment (tbellicha) [inline Application/mobile/app_code/lib/pages/rent/confirm/confirm_rent_state.dart:177]: Bad Key

**Review by tbellicha — CHANGES_REQUESTED** [review]

**Review by invpuppet — APPROVED** [review]
good for me

File: Server/routes/Mobile/rent.js [hunk 15]
```diff
@@ -1,10 +1,61 @@
...
+  itemId,
+  price
+) {
+  try {
+    const formattedDate = formatDate(date);
+      const mailOptions = {
+        from: process.env.SMTP_EMAIL,
+        to: email,
+        subject: 'Confirmation de votre location',
+        text: 'Votre location a bien été enregistrée. \nVous avez loué l\'article ' + itemId + ' pour une durée de ' + duration + ' heures le ' + formattedDate + ' dans le conteneur situé à l\'addresse suivante :  ' + address + ', ' + city + '. \nLe prix total est de ' + price + ' euros. Vous pouvez demander une facture en consultant la location.\nMerci de votre confiance.',
```
Comment (H-chauvet) [inline Server/routes/Mobile/rent.js:29]: Adresse et non pas "Addresse"

**Review by H-chauvet — DISMISSED** [review]
Change the word "Addresse" to "Adresse"

**Review by 31Nathan — CHANGES_REQUESTED** [review]

File: Server/invoice/utils.js
Comment (31Nathan) [inline Server/invoice/utils.js:1]: rent utils and utils in the same directory ? Merge the files or rename it correctly

File: Server/routes/Mobile/createInvoice.js [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+      customerInformationTop + 30
+    )
+    .text("Facturé par :", 50, customerInformationTop + 45)
+    .text("Risu", 150, customerInformationTop + 45)
+
+    .font("Helvetica")
+    .text(invoice.shipping.address, 300, customerInformationTop + 15)
+    .text(
+      invoice.shipping.city +
+        //", " +
```
Comment (31Nathan) [inline Server/routes/Mobile/createInvoice.js:70]: Useless comment

File: Server/routes/Mobile/createInvoice.js
Comment (31Nathan) [inline Server/routes/Mobile/createInvoice.js:1]: It is in the routes directory without routes, should be on the invoice one

**Review by 31Nathan — CHANGES_REQUESTED** [review]
Good invoice directory, now use it and name correctly your file

**Review by 31Nathan — APPROVED** [review]

**Pull request merged**